src/main/java/com/aws/greengrass/util/platforms/unix/UnixExec.java

```diff
@@ -0,0 +1,140 @@
+                // path variable, but without using Exec shorthands to avoid initialization
+                // order paradoxes.
+                Process hack =
+                        Runtime.getRuntime().exec(new String[]{"sh", "-c", "echo 'echo $PATH' | grep -E ':[^ ]'"});
+                StringBuilder path = new StringBuilder();
+
+                Thread bg = new Thread(() -> {
+                    try (InputStream in = hack.getInputStream()) {
+                        for (int c = in.read(); c >= 0; c = in.read()) {
+                            path.append((char) c);
+                        }
+                    } catch (Throwable ignore) {
+                    }
+                });
+                bg.start();
+                bg.join(2000);
+                addPathEntries(path.toString().trim());
+                // Ensure some level of sanity
+                ensurePresent("/bin", "/usr/bin", "/sbin", "/usr/sbin");
+            }
+        } catch (Throwable ex) {
+            staticLogger.atError().log("Error while initializing PATH", ex);
+        }
+        computePathString();
+    }
+
+    private static String deTilde(String s) {
+        if (s.startsWith("~/")) {
+            s = Utils.HOME_PATH.resolve(s.substring(2)).toString();
+        }
+        return s;
+    }
+
+    private static void ensurePresent(String... fns) {
+        for (String fn : fns) {
+            Path ulb = Paths.get(fn);
+            if (Files.isDirectory(ulb) && !paths.contains(ulb)) {
+                paths.add(ulb);
+            }
+        }
+    }
+
+    private static void addPathEntries(String path) {
+        if (path != null && path.length() > 0) {
+            for (String f : path.split("[ :,] *")) {
+                Path p = Paths.get(deTilde(f));
+                if (!paths.contains(p)) {
+                    paths.add(p);
+                }
+            }
+        }
+    }
+
+    @Nullable
+    @Override
+    public Path which(String fn) {
+        fn = deTilde(fn);
+        if (fn.startsWith("/")) {
+            Path f = Paths.get(fn);
+            return Files.isExecutable(f) ? f : null;
+        }
+        for (Path d : paths) {
+            Path f = d.resolve(fn);
+            if (Files.isExecutable(f)) {
+                return f;
+            }
+        }
+        return null;
+    }
+
+    @Override
+    public Optional<Integer> exec() throws InterruptedException, IOException  {
+        // Don't run anything if the current thread is currently interrupted
+        if (Thread.currentThread().isInterrupted()) {
+            logger.atWarn().kv("command", this).log("Refusing to execute because the active thread is interrupted");
+            throw new InterruptedException();
+        }
+        final String[] command = getCommand();
+        logger.atTrace().kv("command", (Supplier<String>) () -> String.join(" ", command)).log();
+        process = Runtime.getRuntime().exec(command, environment, dir);
```

Comment:
While we're making changes, we should use ProcessBuilder instead of runtime exec